database/schema.sql

```diff
@@ -0,0 +1,167 @@
+--
+-- Also, sqlite is said to create indices on PKs automatically, so I will omit them.
+--
+-- As the sqlite will create additional `rowid` unique PK even if another
+--  non integer or integer desc PK is present, some of tables are created
+--  WITHOUT ROWID.
+
+-- Creating 'Courses' table.
+--
+create table if not exists Courses (
+
+    -- Sqlite3 will force ascending primary key to be non-null, replacing
+    --  null with autoincremented key (while sqlite2 won't).
+    --
+    id    INTEGER,
+    desc  TEXT     null,
+
+    primary key (id asc)
+);
+
+-- Creating 'Subjects' table.
+--
+create table if not exists Subjects (
+    id         INTEGER  not null,
+    course_id  INTEGER  not null,
+    desc       TEXT     null,
+
+    primary key (id, course_id),
+    foreign key (course_id) references Courses (id)
+);
+
+create index if not exists Subject_course_ids on Subjects (course_id);
+
+-- Creating 'Students' table.
+--
+create table if not exists Students (
+    -- The "primary key" doesn't make field non null in sqlite.
+    --
+    addr  BLOB  not null,
+
+    primary key (addr)
+
+) without rowid;
+
+-- Creating 'StudentCourses' table.
+--
+-- [Appears to be] many-to-many ref between Students and Courses.
+--
+create table if not exists StudentCourses (
+    student_addr  BLOB     not null,
+    course_id     INTEGER  not null,
+
+    -- Composite primary.
+    --
+    primary key (student_addr, course_id),
+
+    foreign key (student_addr) references Students (addr),
+    foreign key (course_id)    references Courses  (id)
+
+) without rowid;
+
+create index if not exists StudentCourses_student_addr on StudentCourses (student_addr);
+create index if not exists StudentCourses_course_id    on StudentCourses (course_id);
+
+-- Creating 'Assignments' table.
+--
+create table if not exists Assignments (
+    hash           BLOB     not null,
+    course_id      INTEGER  not null,
+    contents_hash  BLOB     not null,
+    desc           TEXT     null,
+
+    primary key (hash),
+    foreign key (course_id) references Courses(id)
+
+) without rowid;
+
+create index if not exists Assigments_course_id on Assignments (course_id);
+
+-- Creating 'StudentAssignments' table.
+--
+-- [Appears to be] many-to-many ref between Students and Assigments.
+--
+create table if not exists StudentAssignments (
+    student_addr     BLOB  not null,
+    assignment_hash  BLOB  not null,
+
+    primary key (student_addr, assignment_hash),
+
+    foreign key (student_addr)    references Students    (addr)
+    foreign key (assignment_hash) references Assignments (hash)
+
+) without rowid;
+
+create index if not exists StudentAssigments_student_addr    on StudentAssignments (student_addr);
+create index if not exists StudentAssigments_assignment_hash on StudentAssignments (assignment_hash);
+
+-- Creating 'Submissions' table.
+--
+create table if not exists Submissions (
+    hash             BLOB     not null,
```

Comment:
This hash is the hash of the other fields right?

Reply:
@lambda76 right